augment evaluate to extract within objects

alexlamsl opened the pull request:
This replaces and extends the existing `"string".length => 6` optimisation, and is also hidden behind `unsafe` compress option flag.

Basically this enables `evaluate` to navigate into object property on a best effort basis, and bails out if the final form isn't constant.

Allows for constants hidden inside configuration objects to propagate, which is often encountered in the field. Does not address #1416 directly but may form part of the final solution.

avdg commented:
I don't see any reason to have this optimization under safe flag on long term if the feature is well maintained and done on literals only.

alexlamsl commented:
@avdg I concur - though since `"string".length` was behind `unsafe`, I erred on the safe side. 😛

avdg commented:
I think it's worth doing an investigation on that if time permits.

alexlamsl commented:
@avdg let's leave that for another PR, as I need incremental understanding on all the corner cases before I'm comfortable enough to tackle the general case 😅

avdg commented:
No worries patch is fine as is and understanding risk takes time. So I understand the precaution.

avdg commented:
The "string".length thingy came from https://github.com/mishoo/UglifyJS2/commit/8511e80f489b4b7e3de8122009d492123a314ca8

kzc commented on lib/compress.js:
```diff
@@ -1117,10 +1134,11 @@ merge(Compressor.prototype, {
-                    return str.length;
+            if (compressor.option("unsafe")) {
+                var val = ev(this.expression, compressor);
+                if (HOP(val, this.property)) {
```
Before calling `HOP()` must check if first argument is an object literal:
```
$ echo 'a = null.p;' | bin/uglifyjs -c unsafe
undefined:6244
                if (ex !== def) throw ex;
                                ^

TypeError: Cannot convert undefined or null to object
    at hasOwnProperty (<anonymous>)
```
We want to preserve bad javascript code as is. Garbage in, garbage out.

alexlamsl replied:
Hmm, your example seems to work over here: 😓 
```
C:\UglifyJS2>echo a = null.p; | node bin\uglifyjs -c unsafe
a=null.p;


C:\UglifyJS2>node
> var minify = require('uglify-js').minify;
undefined
> minify('a = null.p;', { fromString: true, compress: { unsafe: true } });
{ code: 'a=null.p;', map: null }
```

Regardless, do you suggest I follow [this line](https://github.com/mishoo/UglifyJS2/blob/master/lib/compress.js#L2621) to fix this up?

alexlamsl replied:
@kzc fixed, added some tests for that as well

kzc replied:
It just shows that hasOwnProperty is implemented differently on different node versions or platforms.

I suggest something like this:
```
if (val instanceof AST_Object && HOP(val, this.property)) {
```

kzc replied:
Sorry, `val instanceof AST_Object` doesn't work here.

alexlamsl replied:
@kzc except `kzc` is an actual plain object at this point, not an `AST_Node`?

alexlamsl replied:
Just to be clear - does my latest commit address your concern? 😅

kzc replied:
Maybe this is sufficient:
```
if (val && HOP(val, this.property)) {
```

kzc replied:
I think that works.

alexlamsl replied:
@kzc done in https://github.com/mishoo/UglifyJS2/pull/1425/commits/b363786d8b5b7d815197e56b68eebd8058cfcc9b - thanks!

kzc commented:
@avdg If this PR is ever merged to harmony, then an `ev()` would have to be done on the key expression for computed properties for side effects. Not an issue in ES5/master as key is side effect free AFAIK.

avdg commented:
uh yeah, if we get more of these patches that behave totally different on harmony, it would be better that I would have repo access, so I can quickly fix the incompatibles, I don't think @rvanvelzen would be able to do this alone, though I don't want to do this alone myself (as it impacts user code in the end).

For now I might want to delay this pr until I finish my other pr's I'm currently working on, so we have patches for harmony as well.

alexlamsl commented:
@avdg @kzc if you are concerned with `harmony`, I had a quick look and I think if I do `if (prop instanceof AST_ObjectKeyVal && typeof prop.key === "string") { ... }` that should make it safe for both branches?

avdg commented:
I don't know if we should get any issues for harmony actually, but I'm working on something very related (AST_ArrayElement) so I might trip some wires and complicate code merging (but I might also not create these kind of conflicts).

Concerning computed properties, these are formatted `var foo = {['foo' + 'bar']: "foobar!"}` but I don't worry about optimizing these cases. The worst case would be if that code simply doesn't work anymore with any optimization patches.

That said, we might need to keep track of optimizations that have to be implemented on harmony later. I do keep a list for few issues myself, but that's a list only I can access to as that is the project board on https://github.com/avdg/UglifyJS2/projects/

kzc commented:
> I think if I do `if (prop instanceof AST_ObjectKeyVal && typeof prop.key === "string") { ... }` that should make it safe for both branches?

Is `typeof prop.key === "string"` sufficient to avoid computed properties? We might want to use this on master anyway.

alexlamsl commented:
@kzc that would be my understanding, as anything other than a constant key would be an `AST_Node`

kzc commented:
Just to clarify my comment - in `harmony` is `prop` always guaranteed to be of type `AST_ObjectKeyVal`?

avdg commented:
I don't think the ast documentation is up to date for `AST_Object`

kzc commented:
on `harmony`:
```
var AST_ObjectProperty = DEFNODE("ObjectProperty", "key value", {
    $documentation: "Base class for literal object properties",
    $propdoc: {
        key: "[string|AST_Node] the property name converted to a string for ObjectKeyVal. For setters, getters and computed property this is an arbitrary AST_Node",
```
Just `if (typeof prop.key === "string")` should be sufficient I think.

kzc commented:
If a computed prop is a constant expression reduced to a string literal in compress, then `typeof prop.key === "string"` would still work on a subsequent compress pass would it not?

alexlamsl commented:
@kzc updated `prop.key` check as per request

kzc commented on lib/compress.js:
```diff
@@ -1053,7 +1053,7 @@ merge(Compressor.prototype, {
             for (var i = this.properties.length; --i >= 0;) {
                 var prop = this.properties[i];
-                if (prop instanceof AST_ObjectKeyVal) {
+                if (typeof prop.key === "string") {
```
I don't think that's advisable because it could skip a side effect for key/value in a computed prop or a situation where there's a non-string literal key.

kzc replied:
On both master and harmony `ev()` always to be evaluated for `prop.value` in case it throws.

alexlamsl replied:
@kzc would you mind giving me an example of what you mean? 😅

kzc replied:
prop.key can be a Number literal for example.

Let's not worry about `harmony` for this PR. It complicates things.

alexlamsl replied:
Within this loop I'm constructing an object which I can evaluate the values of, filtering out all the others.

The idea is to keep the evaluation process forward, and then at the end see if we hit anything constant. If we don't (or there isn't a key match due to its leaf fails to evaluate), we just bail out.

Now that I'm thinking out loud, I think try-catch-ing the `ev(prop.value, compressor);` here would make sense as I want to skip over those failures.

alexlamsl replied:
@kzc ah, good point - I'll revert back to checking `AST_ObjectKeyVal` then.

kzc replied:
>  I think try-catch-ing the ev(prop.value, compressor); here would make sense as I want to skip over those failures.

No, you must not catch those exceptions - they indicate a side effect and must be allowed to be thrown.

ev() always has to be evaluated on the prop.value regardless of the key.

For ES5/master the key is always constant AFAIK, so ev() does not have to be run on the key. But in harmony (if supported) then ev() must always be run on AST_Node prop.key values in case it throws.

kzc replied:
> I'll revert back to checking AST_ObjectKeyVal then

I don't think that check is needed at all.

alexlamsl replied:
@kzc thanks for bearing with me - I think I get your point now.

Can you check if the tests in the latest commit is good enough to cover this path?

kzc commented on lib/compress.js:
```diff
@@ -1053,9 +1053,7 @@ merge(Compressor.prototype, {
-                if (prop instanceof AST_ObjectKeyVal) {
-                    val[prop.key] = ev(prop.value, compressor);
-                }
+                val[prop.key] = ev(prop.value, compressor);
```
I think that's the correct way to do it for ES5/master.

Can you please add a comment above this live that harmony/ES6 will additionally have to run ev() on prop.key which will throw on side effects?

kzc replied:
@alexlamsl Will the following line work in ES5/master?
```
val[ev(prop.key, compressor)] = ev(prop.value, compressor);
```
If so, maybe that's the answer for both branches.

kzc replied:
Sorry for the the stream of thoughts. Just thinking out loud. Implementing this optimization is more complicated than it first appears.

alexlamsl replied:
@kzc not directly - but I can do:
```js
var key = prop.key;
if (key instanceof AST_Node) {
    key = ev(key, compressor);
}
val[key] = ev(prop.value, compressor);
```

would that be acceptable?

alexlamsl replied:
@kzc no worries! I set aside this weekend to tackle this. I run into this case often enough to want to work on it. 😃

kzc commented:
@alexlamsl I think your "ensure compatibility with harmony branch" fix is okay, but I found a new problem...

expected:
```
$ echo 'console.log({a:1, a:2}.a);' | node
2
```
without your PR:
```
$ echo 'console.log({a:1, a:2}.a);' | uglifyjs -c unsafe | node
2
```
with your PR:
```
$ echo 'console.log({a:1, a:2}.a);' | bin/uglifyjs -c unsafe | node
1
```
Could you please prepend [WIP] to this issue's title while we mull over other potential issues?

alexlamsl commented:
@kzc fix pushed, title updated.

kzc commented:
@alexlamsl Could you please add a few tests with integer and floating point keys - with and without side effects in the value? A key of `0` in particular may prove interesting.

kzc commented on test/compress/evaluate.js:
```diff
@@ -95,6 +101,12 @@ unsafe: {
+            ({a:{b:1},a:1}) + 1,
+            2,
+            ({a:{b:1},a:1}).b + 1,
+            1..b + 1
```
Is `1..b` the correct expected result?

Hard to compare the input with expected because of the length of the tests. Would you mind splitting all the various types of tests into separate named tests - each testing one particular facet?  i.e., object_literal_repeated_props, prop_on_null, object_literal_with_side_effect_in_prop_value, etc.

alexlamsl replied:
@kzc indeed this is getting incomprehensible - I'm splitting them out as we speak 😅 

`1..b` is correct as far as I can tell - one dot would cause syntax error as it confuses with numeral literals.

kzc replied:
Cool. Thanks.

kzc replied:
The test names I gave are just examples - please use whatever test names you think are appropriate.

kzc commented on lib/compress.js:
```diff
@@ -1051,7 +1051,7 @@ merge(Compressor.prototype, {
         def(AST_Object, function(compressor){
             var val = {};
-            for (var i = this.properties.length; --i >= 0;) {
+            for (var i = 0, l = this.properties.length; i < l; i++) {
```
This is a personal preference I admit, but I'd prefer `l` not be used as a variable name because it looks like `1`. Could you please use `len` instead?

alexlamsl commented:
@kzc nice suggestion for trying out numbers as property keys - as they cannot be `AST_Dot`, this exposes a previous issue with `evaluate()` not optimising for `AST_Sub`.

I suppose this wasn't an issue with ES5 as things like `"..."["len" + "gth"]` will eventually be optimised into `"...".length` and feed into `evaluate()` all the same.

Also split out the tests, added new ones, and refactored the eye-sore variable as per request.

kzc commented:
> this exposes a previous issue with evaluate() not optimising for AST_Sub.

Yeah, I just happened to coincidentally notice that during testing this PR as well.  ;-)

The tests are much clearer now. Thanks.

I don't have the time to further review this PR right now unfortunately. Sometimes better to wait on a complex PR like this. Other unforeseen cases usually come up.

alexlamsl commented:
@kzc I'm testing this PR on `node-forge` and has discovered at least one bug to do with LHS expressions.

Thank you for your time so far. If you have any more ideas about what to test, please let me know 😉

kzc commented:
@alexlamsl Happy to review your work. It would be a useful feature.

Since this PR involves code changes that run whether or not the "unsafe" compress option is in effect, we have to be careful.

kzc commented on lib/compress.js:
```diff
@@ -1042,6 +1049,18 @@ merge(Compressor.prototype, {
         def(AST_Constant, function(){
             return this.getValue();
         });
+        def(AST_Object, function(compressor){
```
Probably need an `if (compressor.option("unsafe")) { ... }` wrapper around this logic similar to `def(AST_PropAccess` which calls `throw def;` if the option is not in effect.

alexlamsl replied:
True - otherwise there will be a performance penalty when `unsafe` is off.

alexlamsl replied:
Done.

alexlamsl commented:
Tested `node-forge` and `angular.js` and both seem to run happily after `minify()`.

kzc commented on lib/compress.js:
```diff
@@ -2893,7 +2920,7 @@ merge(Compressor.prototype, {
             }
         }
-        return self;
+        return self.evaluate(compressor)[0];
```
Could you please add a few tests for this AST_Sub property optimization when `unsafe` is `false`?

Ideally testing when the AST_Sub property can be reduced to:
* an integer
* a string literal with no spaces
* a string literal with a space
* special values like null, undefined, NaN or any other you may think of.

And another test where the AST_Sub prop cannot be reduced due to a side effect

alexlamsl replied:
@kzc are [these tests](https://github.com/mishoo/UglifyJS2/pull/1425/commits/3f5cdc8bc8df03a4fe68f8d1e242f8203007b883) what you have in mind?

kzc replied:
@alexlamsl You can keep those new tests but the `sub_properties` tests you've added would still produce the same result without this PR. I'm looking for AST_Sub cases with `-c unsafe=false` that produce different results with and without this PR. I could have sworn that I had created such a case, but I can't seem to recall or manufacture one now.

alexlamsl replied:
@kzc correct me if I'm wrong, but without `unsafe` any `AST_PropAccess._eval(compressor)` will just hit `throw def;` straight away, so `AST_Sub.evaluate(compressor)` should always be a no-op when `unsafe` is disabled.

Before this change, `AST_Sub.evaluate(compressor)` is always a no-op _regardless_ of `unsafe`, so calling it or not makes no difference, apart from may be better performance.

kzc replied:
@alexlamsl Your grasp of the evaluate code is better than mine. What you write makes sense. I clearly wasn't testing what I thought I was testing.

kzc commented:
The use of the new function `isModified` appears to be called outside of the compress "unsafe" option. This seems risky to me as it might introduce an unintended new behavior. Can an "unsafe" check be added to either the function itself or the caller of it?

alexlamsl commented:
@kzc that part of `scope.js` sets a flag which only `reduce_vars` uses, and the only difference is it will now handle the extra cases of `a.b = 1`, i.e. marking `a` as "modified" whereas it didn't before.

`isModified()` is a read-only function. Moreover, I don't think we have access to `options.compress` thus the `unsafe` flag [within `figure_out_scope()`](https://github.com/mishoo/UglifyJS2/blob/master/tools/node.js#L97)

kzc commented:
I see. What's the worst case consequence of marking something as "modified" as opposed to not marking it?

It's true that options are not readily accessible from different modules - which is why eventually I'd like to see a single flattened options object for all functionality. Out of scope for this PR - UglifyJS3 perhaps.

kzc commented:
Regarding figure_out_scope() in minify() - there's a known bug there:

Have minify() and tests use figure_out_scope() as uglifyjs CLI does #1408

alexlamsl commented:
@kzc setting the "modified" flag prohibits `evaluate()` from [value substitution](https://github.com/mishoo/UglifyJS2/blob/master/lib/compress.js#L1111), which would make for less optimised output

alexlamsl commented:
Yes I am aware of #1408 - it only passes in `options.mangle`, not `options.compress`, so won't make any difference in this case.

kzc commented:
>  setting the "modified" flag prohibits evaluate() from value substitution, which would make for less optimized output

So worst case scenario for misuse is just sub-optimal code?

kzc commented:
> Yes I am aware of #1408 - it only passes in options.mangle, not options.compress, so won't make any difference in this case.

Groan. I see. Yet another reason why we need a single flat options object.

At least minify() and bin/uglifyjs will behave the same in that respect.

kzc commented:
Cool. Stepping away from computer.

kzc commented:
With a little work this PR could be extended to support AST_Array:
```js
    // methods to evaluate a constant expression
    (function (def){
```
...
```js
        def(AST_Array, function(compressor){
            if (compressor.option("unsafe")) {
                var val = [];
                for (var i = 0, len = this.elements.length; i < len; i++) {
                    val[i] = ev(this.elements[i], compressor);
                }
                return val;
            }
            throw def;
        });
```
```
$ echo 'f("abcdef"[2]);' | bin/uglifyjs -c unsafe
f("c");
```
```
$ echo 'f([10,20,30][2-1]);' | bin/uglifyjs -c unsafe
f(20);
```

alexlamsl commented:
@kzc just woke up and need to take care of day business - I'll be back with array optimisation this evening 😉

alexlamsl commented:
@kzc added relevant tests for index and `.length` optimisations

alexlamsl commented on test/compress/arrays.js:
```diff
@@ -72,3 +72,39 @@ constant_join_2: {
+
+        function f1() {
+            var a = [1, 2, 3];
+            for (var i = 0, len = 3; i < len; i++)
```
TODO: investigate why `i < len` isn't being optimised by `evaluate`

kzc commented on test/compress/evaluate.js:
```diff
@@ -37,3 +37,347 @@ positive_zero: {
+            [1, 2, 3, 4][0] + 1,
+            [1, 2, 3, 4][6 - 5] + 1,
+            [[1, 2], [3, 4]][0] + 1,
+            [[1, 2], [3, 4]][6 - 5][1] + 1
```
Could you add a test for an array hole? It should equate to `void 0`.
```
[10, , 30][1]
```

alexlamsl replied:
Added two cases for holes.

kzc commented:
There's an issue with the interaction of reduce_vars and unsafe:

expected:
```
$ echo 'var a = [1], a = [2]; console.log(a === a);' | node
true
```

actual:
```
$ echo 'var a = [1], a = [2]; console.log(a === a);' | bin/uglifyjs -c reduce_vars,unsafe 
var a=[1],a=[2];console.log(!1);
```
```
$ echo 'var a = [1], a = [2]; console.log(a === a);' | bin/uglifyjs -c reduce_vars,unsafe | node
false
```
Same issue with objects:

expected:
```
$ echo 'var a = {x:1}, a = {x:1}; console.log(a === a);' | node
true
```
actual:
```
$ echo 'var a = {x:1}, a = {x:1}; console.log(a === a);' | bin/uglifyjs -c reduce_vars,unsafe | node
false
```

alexlamsl commented:
@kzc wow that looks exotic. Thanks for the finding - working on it!

kzc commented on test/compress/evaluate.js:
```diff
@@ -37,3 +37,351 @@ positive_zero: {
+            NaN,
+            "1,21",
+            5,
+            (void 0)[1] + 1
```
Could you please add a additional simpler test for an array hole that just results in `void 0`? 
Something like: `[10,,30][1]`

alexlamsl replied:
Done - sorry for the over-complication 😅

kzc commented:
> wow that looks exotic. 

I was concerned about object identity and the now constant object/array literals under `unsafe`. There may be other issues related to this.

alexlamsl commented:
FWIW, `var a = {}; console.log(a === a);` would trigger the issue. Aiming to just cache the `_eval()` results on these `AST_SymbolRef` nodes, which should also provide some performance benefits.

kzc commented:
Here's a variation of the same object identity issue that's more likely to happen in real life code:

expected:
```
$ echo 'var a = [1], b = a, c = a; console.log(b === c);' | node
true
```

actual:
```
$ echo 'var a = [1], b = a, c = a; console.log(b === c);' | bin/uglifyjs -c reduce_vars,unsafe | node
false
```

kzc commented:
> Aiming to just cache the _eval() results on these AST_SymbolRef nodes

What's your strategy? Adding a new field would have an adverse impact on other transforms.

alexlamsl commented:
I'm thinking of [replacing `ev(d.init, compressor);` with](https://github.com/mishoo/UglifyJS2/pull/1425/files#diff-230c53e13e3ee367aec7d49470eb42caL1112):
```js
if (!HOP(d.init, '_evaluated')) {
    d.init._evaluated = ev(d.init, compressor);
}
return d.init._evaluated;
```
Is this too intrusive? The alternative is to have a local look-up for every `evaluate()`, i.e. the result(s) will not be reused across invocations.

alexlamsl commented:
@kzc included https://github.com/mishoo/UglifyJS2/pull/1425#issuecomment-272879828 in the latest commit - will now run through another iteration of `forge.js` & `angular.js` for potential new ways to crash and burn 😎

kzc commented:
Regarding the d.init._evaluated fix, there's a concern that there might now be two sources of truth which could diverge with other transforms.

alexlamsl commented:
@kzc I share your concern. I'm fine with confining the cache as a double-array instead. The question is, should it be within `evaluate()`, or are we safe enough to cache across the `compressor` object?

alexlamsl commented:
@kzc went ahead with `evaluate()` caching, which seems to pass all existing tests 🤞

kzc commented:
@alexlamsl Your `d.init._evaluated` solution might be okay. Just thinking out loud... `d.init._evaluated` is only set and used if the symbol definition is a constant. That limits the types of transforms that could potentially be done on the node. But you'd have to look how `.init` is used throughout compress.js to see if it's problematic.

Regarding the evaluate caching - not crazy about the linear lookups with indexOf on all symbol refs. I think it would slow down uglification noticeably on large source inputs. This could easily be tested and confirmed or refuted.

In either case whether `d.init._evaluated` or evaluate caching is used, it would have to be cleared in `clear_opt_flags` for the `-c passes=N` option.

kzc commented:
I think we need an object identity test where the same "constant" object travels through two different functions to see if they still match upon return.

alexlamsl commented:
@kzc that last commit is just me thinking out loud. I think it's way too intrusive to solve a relatively smaller issue.

I will stick back with `d.init._evaluated` and will clear it out in `clear_opt_flags()` as you stated. 😉

kzc commented:
I think that's a good idea.

Can we limit `d.init._evaluated` to only be used when `unsafe` is `true` to limit the potential impact?  It might be fine in the general case, but I can't be 100% certain.

alexlamsl commented:
`reduce_vars` without `unsafe` [is left untouched by this PR now](https://github.com/mishoo/UglifyJS2/pull/1425/files#diff-230c53e13e3ee367aec7d49470eb42caR1154)

kzc commented:
Very good. Thanks.

Did you think this PR would be as complicated as it is?

alexlamsl commented:
@kzc this is where I stopped last time when I implemented `reduce_vars`, so I did expect this to get real hairy.

To be fair, the complication is with taking care of all the corner cases - the actual code changes are fairly minimal.

alexlamsl commented:
Next PR(s) would be to take care of those `OPT()` cases which don't call `evaluate()` one reason or another, just like the one you spotted for `AST_Sub`.

I see things like `a < b` and `console.log(a)` not being optimised, for instance.

kzc commented:
> the complication is with taking care of all the corner cases - the actual code changes are fairly minimal

No doubt. Javascript is a big collection of corner cases.

kzc commented:
> `a < b`

Be warned - there be dragons in Javascript inequality statements. Talk about corner cases.

alexlamsl commented:
By `a < b` I mean turning `var a = 3; console.log(a < b)` to `console.log(3 < b)` - I hope that's safe? 👻

alexlamsl commented:
Anyway, give me a shout when new error cases are found and I'll work on them some more. Otherwise I'll leave this PR here for you to merge at your leisure 😎

avdg commented:
I guess that specific example is fine, but variable access can have weird stuff installed (rarely used though). Things like https://github.com/tc39/test262/blob/master/test/language/expressions/less-than-or-equal/S11.8.3_A2.3_T1.js

kzc commented:
@alexlamsl I think this PR is in good shape but it would be wise to wait a week before removing the [WIP] just in case someone finds some breakage.

alexlamsl commented:
@avdg thanks for the pointer - I'll try to keep that in mind for my next adventure!

@kzc sounds like a good plan.

avdg commented:
[wip] just means that there are planned changes to take into account. If there is nothing planned, [wip] isn't necessary. There is still a lot that can go wrong before (and sadly enough just after) merging.

kzc commented:
Regarding inequality comparisons, see `unsafe_comps` in compress.js.

kzc commented:
> If there is nothing planned, [wip] isn't necessary.

I personally take the bug report frequency into account before removing [WIP] on complicated far-reaching enhancements. If no reports for a while, it's usually safer.

kzc commented:
@alexlamsl Here's a new bug in code I found in the wild:
```
$ echo 'Object({toString: 0} + "");' | bin/uglifyjs -c unsafe=true
undefined:6259
                if (ex !== def) throw ex;
                                ^

TypeError: Cannot convert object to primitive value
```

alexlamsl commented:
@kzc interesting hack - fixed 😉

kzc commented:
Assuming no additional bugs are found please squash PR into a single commit when [WIP] is removed. Thanks.

alexlamsl commented:
rebased & squashed

rvanvelzen commented:
This is looking very good, thanks a lot for the hard work. :)

alexlamsl commented:
If I'm correct this should resolve #787
